### PR TITLE
New_subtract_two_numbers

### PR DESCRIPTION
**Pull request by GeehanAli**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in any docstrings or tests
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/subtract_two_numbers.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# python3
+# coding
+"""creating on 01 03 2025
+@author Geehan Ali + ChatGPT(phind)"""
+
+from typing import Union
+
+
+def subtract_two_numbers(
+    num1: Union[int, float], num2: Union[int, float]
```
Comment (misik-eng) [inline solutions/subtract_two_numbers.py:10]: Nice work,  I see that you changed the variable names, looks good.

File: solutions/subtract_two_numbers.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        >>> subtract_two_numbers(-5.5, 2)
+        -7.5
+        >>> subtract_two_numbers(-5.5, -4.5)
+        -1
+
+    Raises:
+        TypeError: If either num1 or num2 is not a number (int or float).
+
+    """
+    assert isinstance(
```
Comment (misik-eng) [inline solutions/subtract_two_numbers.py:40]: Maybe you could have used a single line for assertions to make it more readable, and you could use 'TypeError', but this also works. Thank you.

File: solutions/tests/test_subtract_two_numbers.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+
+def test_float_and_large_numbers(self):
+    """Test subtraction with floating-point numbers and large integers."""
+    self.assertAlmostEqual(subtract_two_numbers(0.001, 0.0009), 0.00010000000000000005)
+
+
+def test_defensive_case_non_numeric_input(self):
+    """Test handling of non-numeric input."""
+    with self.assertRaises(AssertionError):
```
Comment (misik-eng) [inline solutions/tests/test_subtract_two_numbers.py:49]: You could have used TypeError, but it's not strictly necessary. This is also looks good. Thank you.

File: solutions/tests/test_subtract_two_numbers.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import unittest
+
+from ..subtract_two_numbers import subtract_two_numbers  # type: ignore
+
+
+class TestSubtractTwoNumbers(unittest.TestCase):
+    """Test suite for subtract_two_numbers function"""
+
+    def test_normal_subtraction(self):
+        """Test normal subtraction of two numbers."""
```
Comment (misik-eng) [inline solutions/tests/test_subtract_two_numbers.py:18]: I see that you added a clear explanation for each tests, looks good!

File: solutions/subtract_two_numbers.py (hunk 1 above)
Comment (GeehanAli) [reply solutions/subtract_two_numbers.py:10]: Yes, I did 
Thanks!

File: solutions/tests/test_subtract_two_numbers.py (hunk 4 above)
Comment (GeehanAli) [reply solutions/tests/test_subtract_two_numbers.py:18]: Yes,
I fixed it
Thanks!

File: solutions/subtract_two_numbers.py (hunk 1 above)
Comment (GeehanAli) [reply solutions/subtract_two_numbers.py:10]: Yes, I did.
Thanks!

File: solutions/tests/test_subtract_two_numbers.py (hunk 4 above)
Comment (GeehanAli) [reply solutions/tests/test_subtract_two_numbers.py:18]: Yes, I updated it to match the checklist. Thank you for pointing that out!

**Review by misik-eng — APPROVED** [review]
I believe You could have done it with one branch but sometimes it is easier to create a new one and start from clear page, I know , hehe , It looks awesome and clear.

One more thing I'd like to mention: Gennadii suggested committing more frequently during development, and I think it's a great idea. This approach can really help with tracking progress and reviewing changes step by step.

But your challenge looks good,    Approved
### **Files**

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name - `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

---

### **Unit Tests**

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has:
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
  - **Comment:** You have defensive assertions to handle non-numeric inputs correctly.
- [x] There are tests for boundary cases
  - **Comment:** Boundary cases like large numbers and zero are well-tested.

---

### **Function Docstring**

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
- The return value is described:
  - [x] Type
- [x] The defensive assertions are documented using `Raises:`
  - **Comment:** Consider specifying the exact error raised (e.g., `TypeError`) in the docstring.
- [x] Include 3 or more (passing!) doctests
  - **Comment:** Great job including multiple doctests that validate expected behavior.

---

### **The Function**

- [x] The function's name describes its behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file

---

## **Strategy**

### **Do's**

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

---

## **Implementation**

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
- [x] Variables are named with snake_case
  - **Comment:** Variable names are consistent with PEP 8.
- [x] The code follows the strategy as simply as possible
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
  - **Comment:** Defensive checks for argument types are appropriate.
- [x] Defensive assertions include as little logic as possible

File: solutions/subtract_two_numbers.py (hunk 2 above)
Comment (misik-eng) [reply solutions/subtract_two_numbers.py:40]: It looks great Just I thought that you could use just one line like    
   `assert isinstance( num1, (int, float)), "First argument must be a number (int or float)"`

File: solutions/subtract_two_numbers.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        -20
+        >>> subtract_two_numbers(-12.5, -4)
+        -8.5
+        >>> subtract_two_numbers(-5.5, 2)
+        -7.5
+        >>> subtract_two_numbers(-5.5, -4.5)
+        -1
+
+    Raises:
+        TypeError: If either num1 or num2 is not a number (int or float).
```
Comment (misik-eng) [inline solutions/subtract_two_numbers.py:37]: I couldn't see any TypeError but it is kinda a similar so  AssertioErrors is also good
